practices/unit_tests.md

```diff
@@ -0,0 +1,93 @@
+---
+slug: unit-tests
+aspect: processes
+stages:
+- development
+short_description: Unit tests are a foundation of your testing. They test the smallest pieces of code called units that have to be isolated from a system. Unit tests are small, cheap and easy to maintain. The number of unit tests outnumbers any other types of tests.
+tags:
+- unit
+- test
+- function
+- code
+- coverage
+- behavior
+keywords:
+- units
+- code
+- isolation
+- test pyramid
+- arrange
+- act
+- assert
+---
+
+# Unit Tests
+
+## What Are the Unit Tests
+
+Unit testing is a way to test units - the smallest components of your software, the smallest piece of code. A unit can be a single function. The goal is to validate that each unit performs as it should. A unit test tests a behavior in isolation to other tests. If the test relies on an external system, it is not a unit test. Unit tests should be written during the design phase, prior to implementation to prevent defects from being deployed to production. They should be run every time the code is changed and provide a clear description of the feature being tested.
+
+Unit test coverage is a subtype of [code coverage](/issues/code-coverage) - it shows which lines of code were tested by unit tests. However, the goal is not to have a 100% unit test coverage. In fact, you should have as few unit tests as possible and cover as many types of behavior as possible. 100% unit test coverage exists only in a perfect world. While writing unit tests focus on behaviors - adding new tests just to improve coverage leads to tests without any purpose and you would be just wasting your time.
+
+Unit testing can be done manually but it is usually automated. In terms of automated tests, there is a concept called **the test pyramid** that shows how to efficiently balance the automated tests.
+
+![Test Pyramid](/files/unit_tests.png)
+
```

Comment:
I have a bug in the formatting - please remove this empty line. I'll fix it soon.

Reply:
I don't know how to fix this bug, I tried to rewrite it multiple times. The picture is deleted for now.